File: src/Java/1. First Course/3. Second Hole/src/main/refactoringgolf/store/OrderItem.java
package refactoringgolf.store;

public class OrderItem {

	public Product product;
	public int quantity;

	/*
	 * Order Item Constructor
	 */
	public OrderItem(Product product, int quantity) {
		this.product = product;
		this.quantity = quantity;
	}

	public Product getProduct() {
		return product;
	}

	public int getQuantity() {
		return quantity;
	}

	public float total() {
		float discount = 0;
		if (getProduct().getCategory() == ProductCategory.Accessories) {
			discount = calculateAccessoriesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Bikes) {
			discount = calculateBikesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Cloathing) {
			discount = calculateCloathingDiscount();
		}
		return unitPricePerQuantity() - discount;
	}

	private float calculateAccessoriesDiscount() {
		float discount = 0;
		float unitPricePerQuantity = unitPricePerQuantity();
		if (unitPricePerQuantity >= 100) {
			discount = unitPricePerQuantity * 10 / 100;
		}
		return discount;
	}

	private float calculateBikesDiscount() {
		return unitPricePerQuantity() * 20 / 100;
	}

	private float calculateCloathingDiscount() {
		float discount = 0;
		if (getQuantity() > 2) {
			discount = getProduct().getUnitPrice();
		}
		return discount;
	}

	private float unitPricePerQuantity() {
		return getProduct().getUnitPrice() * getQuantity();
	}
}

File: src/Net/1. First Course/2. Second Hole/RefactoringGolf.Store/OrderItem.cs
namespace RefactoringGolf.Stack
{
    public class OrderItem
    {
        public Product Product { get; private set; }

        public int Quantity { get; private set; }

        /*
         * Order Item Constructor
         */
        public OrderItem(Product product, int quantity)
        {
            Product = product;
            Quantity = quantity;
        }

        public decimal Total()
        {
            decimal discount = 0;
            if (Product.Category == ProductCategory.Accessories)
            {
                discount = this.CalculateAccessoriesDiscount();
            }
            if (Product.Category == ProductCategory.Bikes)
            {
                discount = this.CalculateBikesDiscount();
            }
            if (Product.Category == ProductCategory.Cloathing)
            {
                discount = this.CalculateCloathingDiscount();
            }
            return this.ItemAmount() - discount;
        }

        private decimal CalculateAccessoriesDiscount()
        {
            decimal discount = 0;
            if (this.ItemAmount() >= 100)
            {
                discount = this.ItemAmount() * 10 / 100;
            }
            return discount;
        }

        private decimal CalculateBikesDiscount()
        {
            return this.ItemAmount() * 20 / 100;
        }

        private decimal CalculateCloathingDiscount()
        {
            decimal discount = 0;
            if (this.Quantity > 2)
            {
                discount = this.Product.UnitPrice;
            }
            return discount;
        }

        private decimal ItemAmount()
        {
            return this.Product.UnitPrice * this.Quantity;
        }
    }
}